File: leetcode/binary_search/1231_Divide_Chocolate/codes/bs.py
from typing import List

class Solution:
    def maximizeSweetness(self, sweetness: List[int], k: int) -> int:
        def canDivide(minSweet):
            count, total = 0, 0
            for s in sweetness:
                total += s
                if total >= minSweet:
                    count += 1
                    total = 0
            return count >= k + 1

        left, right = 1, sum(sweetness)
        while left < right:
            mid = (left + right + 1) // 2
            if canDivide(mid):
                left = mid
            else:
                right = mid - 1
        return left

File: leetcode/binary_search/1231_Divide_Chocolate/codes/bs.c
#include <stdio.h>

int canDivide(int* sweetness, int n, int k, int minSweet) {
    int count = 0, total = 0;
    for (int i = 0; i < n; ++i) {
        total += sweetness[i];
        if (total >= minSweet) {
            count++;
            total = 0;
        }
    }
    return count >= k + 1;
}

int maximizeSweetness(int* sweetness, int n, int k) {
    int sum = 0;
    for (int i = 0; i < n; ++i) sum += sweetness[i];

    int left = 1, right = sum;
    while (left < right) {
        int mid = (left + right + 1) / 2;
        if (canDivide(sweetness, n, k, mid)) {
            left = mid;
        } else {
            right = mid - 1;
        }
    }
    return left;
}